File: aoc2024/ocaml/lib/day11.ml
let file = Util.get_input_filename 11
(* let file = Util.get_test_filename 11 *)

let to_int (txt : string) : int list =
  String.split_on_char ' ' txt |> List.map int_of_string

let read_file (name : string) : int list =
  let ic = open_in name in
  let try_read () = try Some (input_line ic) with End_of_file -> None in
  let rec loop acc =
    match try_read () with
    | None ->
        close_in ic;
        List.rev acc
    | Some s -> loop (to_int s :: acc)
  in
  loop [] |> List.flatten

let has_even_digits (stone : int) : bool =
  let stone_length = Int.to_string stone |> String.length in
  stone_length mod 2 = 0

let split_stone (stone : int) : int list =
  let str_stone = Int.to_string stone in
  let stone_length = String.length str_stone in
  let new_length = stone_length / 2 in
  let str_stone1 = String.sub str_stone 0 new_length in
  let str_stone2 = String.sub str_stone new_length new_length in
  [ int_of_string str_stone1; int_of_string str_stone2 ]

let apply_rule (stone : int) : int list =
  match stone with
  | 0 -> [ 1 ]
  | s when has_even_digits s -> split_stone s
  | s -> [ 2024 * s ]

let rec stones_after_blink (stone, iter) =
  match iter with
  | i when i <= 0 -> failwith "Invalid iter"
  | 1 -> apply_rule stone |> List.length
  | i -> (
      let gen_stones = apply_rule stone in
      match gen_stones with
      | s :: [] -> stones_after_blink (s, i - 1)
      | [ s1; s2 ] ->
          stones_after_blink (s1, i - 1) + stones_after_blink (s2, i - 1)
      | _ -> failwith "Invalid result")

let stones_after_blink_memo =
  let stones_after_blink self (stone, iter) =
    match iter with
    | i when i <= 0 -> failwith "Invalid iter"
    | 1 -> apply_rule stone |> List.length
    | i -> (
        let gen_stones = apply_rule stone in
        match gen_stones with
        | s :: [] -> self (s, i - 1)
        | [ s1; s2 ] -> self (s1, i - 1) + self (s2, i - 1)
        | _ -> failwith "Invalid result")
  in
  Util.memo_rec stones_after_blink

let get_num_stones stones iter fn =
  List.fold_left (fun acc s -> acc + fn (s, iter)) 0 stones

let part1 () =
  let stones = read_file file in
  let num_stones = get_num_stones stones 25 stones_after_blink in
  Printf.printf "Number of stones for 25: %d\n" num_stones

let part2 () =
  let stones = read_file file in
  let num_stones = get_num_stones stones 75 stones_after_blink_memo in
  Printf.printf "Number of stones for 75: %d\n" num_stones

File: aoc2024/elixir/lib/day11.ex
defmodule AOC.D11 do
  use Memoize

  defp input_file(), do: Util.get_input_filename(11)
  # defp input_file(), do: Util.get_test_filename(11)

  defp to_int(value) do
    {num, _} = Integer.parse(value)
    num
  end

  defp has_even_digits(val) do
    num_digits = Integer.to_string(val) |> String.length()
    rem(num_digits, 2) == 0
  end

  defp split_stone(stone) do
    num_str = Integer.to_string(stone)
    new_length = round((num_str |> String.length()) / 2)
    <<stone1::binary-size(new_length), stone2::binary>> = num_str
    [to_int(stone1), to_int(stone2)]
  end

  defp apply_rule(stone) do
    if stone == 0 do
      [1]
    else
      if has_even_digits(stone) do
        split_stone(stone)
      else
        [2024 * stone]
      end
    end
  end

  defp apply_rules(stones) do
    List.foldl(stones, [], fn s, acc -> apply_rule(s) ++ acc end)
  end

  defp apply_rules(stones, iter, i) do
    IO.write("\r#{Progress.bar(i, iter)}")

    if i < iter do
      stones = apply_rules(stones)
      apply_rules(stones, iter, i + 1)
    else
      IO.write("\n")
      stones
    end
  end

  defp apply_rules(stones, iter) do
    apply_rules(stones, iter, 0)
  end

  defp read_file(file) do
    File.read!(file)
    |> String.split(" ")
    |> Enum.filter(fn s -> s != "" end)
    |> Enum.map(fn s -> to_int(s) end)
  end

  def part1 do
    stones =
      input_file() |> read_file()

    num_blinking = 25
    num_stones = apply_rules(stones, num_blinking) |> length()
    IO.puts("Number of stones for #{num_blinking}: #{num_stones}")
  end

  defmemo stones_after_blink(_stone, iter) when iter <= 0 do
    raise "Invalid iter"
  end

  defmemo stones_after_blink(stone, 1) do
    gen_stones = apply_rule(stone)
    length(gen_stones)
  end

  defmemo stones_after_blink(stone, iter) do
    gen_stones = apply_rule(stone)

    case gen_stones do
      [stone] ->
        stones_after_blink(stone, iter - 1)

      [stone1, stone2] ->
        stones_after_blink(stone1, iter - 1) + stones_after_blink(stone2, iter - 1)

      _ ->
        raise "Invalid result"
    end
  end

  defp get_num_stones(stones, iter) do
    List.foldl(stones, 0, fn s, acc -> stones_after_blink(s, iter) + acc end)
  end

  def part2 do
    stones =
      input_file() |> read_file()

    num_blinking = 75
    # num_blinking = 25
    num_stones = get_num_stones(stones, num_blinking)
    IO.puts("Number of stones for #{num_blinking}: #{num_stones}")
  end

  def test_progress do
    total = 50

    Enum.each(1..total, fn task ->
      IO.write("\r#{Progress.bar(task, total)}")
      Process.sleep(50)
    end)

    IO.write("\n")
  end
end
